Experimental ref panel: when clicking definition in mini code view, open code view immediately

mrnugget opened the pull request:
This changes the behaviour of the code blob on the right: it now stays open when the user landed on a reference that they clicked on _inside the code view_.

In order to implement this I had to simplify the component structure a bit, which I think is a good thing.

https://user-images.githubusercontent.com/1185253/156743318-1a2b55e6-b6c7-4b80-a26a-68b174efafdd.mp4

## Test plan

- Tested manually all the different behaviours. See video.

umpox commented on client/web/src/global/CoolCodeIntel.tsx:
```diff
@@ -180,18 +182,64 @@ export const ReferencesList: React.FunctionComponent<
+            // get from hoverifier is 1-indexed.
+            line: props.token.line - 1,
+            character: props.token.character - 1,
+            filter: filter || null,
```
```suggestion
            filter: debouncedFilter || null,
```

Right? We don't want to spam network requests

mrnugget replied:
Yes! Thanks for this, I just noticed that something was off.

umpox commented on client/web/src/global/CoolCodeIntel.tsx:
```diff
@@ -180,18 +182,64 @@ export const ReferencesList: React.FunctionComponent<
+    // definitions) we select the first definition. We set it as activeLocation
+    // and push it to the blobMemoryHistory so the code blob is open.
+    useEffect(() => {
+        if (props.jumpToFirst === true && definitions.length > 0) {
```
```suggestion
        if (props.jumpToFirst && definitions.length > 0) {
```

As it's just `boolean | undefined` can we simplify

umpox commented on client/web/src/global/CoolCodeIntel.tsx:
```diff
@@ -221,17 +298,45 @@ export const ReferencesList: React.FunctionComponent<
                 type="text"
                 placeholder="Filter by filename..."
-                value={filter === undefined ? '' : filter}
+                value={debouncedFilter === undefined ? '' : debouncedFilter}
```
```suggestion
                value={filter === undefined ? '' : filter}
```

We want the input UI to be as responsive as possible, but the actual network request should be the one that is debounced. 

Also, I think we should decouple this from the loading/error/!data states. The problem here will be that the new filter value will trigger a new request, render a loading spinner and unmount this input, which will make the element lose focus and maybe lead to other bugs. Easiest solution would probably be to hoist this out of `ReferencesList` and manage the `Input` in the parent component

umpox commented on client/web/src/global/CoolCodeIntel.tsx:
```diff
@@ -207,12 +255,41 @@ export const ReferencesList: React.FunctionComponent<
+        panelHistory.push(appendJumpToFirstQueryParameter(url) + toViewStateHash('references'))
+    }
+
+    if (loading) {
+        return (
+            <>
+                <LoadingSpinner inline={false} className="mx-auto my-4" />
+                <p className="text-muted text-center">
+                    <i>Loading precise code intel ...</i>
+                </p>
+            </>
+        )
+    }
```
As mentioned on Slack, it would be nice to avoid this blocking the whole panel UI. Ideally we can keep the blob open as much as possible for a seamless UX (especially for clicking tokens in the same file)

mrnugget replied:
Done in [45a5ab9](https://github.com/sourcegraph/sourcegraph/pull/32146/commits/45a5ab97bd79e1d9d0a74856dda61581c6bace20)

umpox reviewed (approved):
Looks good, left some comments with a few things I noticed. Nothing too blocking though!

mrnugget commented:
Fixed the filtering, added a mini loading spinner (design still needs to be tweaked) and added optimistic in-file navigation. Feels quite a bit better now:

https://user-images.githubusercontent.com/1185253/156772832-cc98dc2f-69fa-4876-abb7-2e8a1aa217b2.mp4
